cornac/eval_methods/propensity_stratified_evaluation.py

```diff
@@ -0,0 +1,514 @@
+            if v > fmin:
+                item_freq[k] = pow(v, alpha)
+
+        return item_freq  # user-independent propensity estimations
+
+    def _build_stratified_datasets(self, train_data, test_data, val_data):
+
+        if train_data is None or len(train_data) == 0:
+            raise ValueError("train_data is required but None or empty!")
+        if test_data is None or len(test_data) == 0:
+            raise ValueError("test_data is required but None or empty!")
+
+        self.global_uid_map.clear()
+        self.global_iid_map.clear()
+
+        # build training set
+        self.train_set = Dataset.build(
+            data=train_data,
+            fmt=self.fmt,
+            global_uid_map=self.global_uid_map,
+            global_iid_map=self.global_iid_map,
+            seed=self.seed,
+            exclude_unknowns=False,
+        )
+        if self.verbose:
+            print("---")
+            print("Training data:")
+            print("Number of users = {}".format(self.train_set.num_users))
+            print("Number of items = {}".format(self.train_set.num_items))
+            print("Number of ratings = {}".format(self.train_set.num_ratings))
+            print("Max rating = {:.1f}".format(self.train_set.max_rating))
+            print("Min rating = {:.1f}".format(self.train_set.min_rating))
+            print("Global mean = {:.1f}".format(self.train_set.global_mean))
+
+        # build test set
+        self.test_set = Dataset.build(
+            data=test_data,
+            fmt=self.fmt,
+            global_uid_map=self.global_uid_map,
+            global_iid_map=self.global_iid_map,
+            seed=self.seed,
+            exclude_unknowns=self.exclude_unknowns,
+        )
+        if self.verbose:
+            print("---")
+            print("Test data (Q0):")
+            print("Number of users = {}".format(len(self.test_set.uid_map)))
+            print("Number of items = {}".format(len(self.test_set.iid_map)))
+            print("Number of ratings = {}".format(self.test_set.num_ratings))
+            print("Max rating = {:.1f}".format(self.test_set.max_rating))
+            print("Min rating = {:.1f}".format(self.test_set.min_rating))
+            print("Global mean = {:.1f}".format(self.test_set.global_mean))
+            print(
+                "Number of unknown users = {}".format(
+                    self.test_set.num_users - self.train_set.num_users
+                )
+            )
+            print(
+                "Number of unknown items = {}".format(
+                    self.test_set.num_items - self.train_set.num_items
+                )
+            )
+
+        # build stratified datasets
+        self.stratified_sets = {}
+
+        # match the corresponding propensity score for each feedback
+        test_props = np.array([self.props[i]
+                               for u, i, r in test_data], dtype=np.float64)
+
+        # stratify
+        strata, bins = pd.cut(x=test_props,
```

Comment:
Instead of using pandas, we can use numpy for this binning to avoid adding more dependency to the library.

https://numpy.org/doc/stable/reference/generated/numpy.digitize.html